### PR TITLE
Implemented scrolling the editor grid by holding the mouse wheel.

### PR DESCRIPTION
**Pull request by Brollyy**
#102

Once middle button is pressed, the icon appears in that position. Moving the mouse will then scroll the grid automatically.
If the middle button was not released before moving the mouse, releasing it will automatically exit this mode.
If the middle button was released before moving the mouse, this mode can be exited by clicking either left, right or middle button.

### CODE REVIEW

**Comment by Brollyy** [comment]
Inconsistent behavior with the note selection - if user holds the left button before entering this mode, the selection rectangle isn't updated properly when mouse is moved and if left button is released while in this mode, it cancels the scrolling, but the selection mode is still active and can only be cancelled by left click, not selecting any notes.

File: Windows/MainWindow.GridControls.cs [hunk 1]
```diff
@@ -102,20 +106,72 @@ public partial class MainWindow : Window {
...
+                // update beat display
+                lblSelectedBeat.Content = $"Time: {Helper.TimeFormat(gridController.snappedBeat * 60 / globalBPM)}, Global Beat: {Math.Round(gridController.snappedBeat, 3)} ({Math.Round(gridController.unsnappedBeat, 3)})";
 
-            // update beat display
-            lblSelectedBeat.Content = $"Time: {Helper.TimeFormat(gridController.snappedBeat * 60 / globalBPM)}, Global Beat: {Math.Round(gridController.snappedBeat, 3)} ({Math.Round(gridController.unsnappedBeat, 3)})";
+                // initiate drag selection
+                if (editorMouseDown) {
+                    Vector delta = mousePos - editorDragSelectStart;
+                    if (delta.Length > Editor.DragInitThreshold) {
+                        gridController.BeginDragSelection(mousePos);
```
Comment (Brollyy) [inline Windows/MainWindow.GridControls.cs:131]: Inconsistent behavior with the note selection - if user holds the left button before entering this mode, the selection rectangle isn't updated properly when mouse is moved and if left button is released while in this mode, it cancels the scrolling, but the selection mode is still active and can only be cancelled by left click, not selecting any notes.

File: Windows/MainWindow.GridControls.cs [hunk 2]
```diff
@@ -135,24 +189,38 @@ public partial class MainWindow : Window {
             if (Keyboard.FocusedElement is TextBox || !editorIsLoaded) {
                 return;
             }
-            Point mousePos = e.GetPosition(EditorGrid);
-            editorDragSelectStart = mousePos;
-            gridController.GridMouseDown(mousePos);
-            editorMouseDown = true;
+            if (editorIsHoldScrolling) { //Moving with a released wheel and pressing a button
+                scrollEditor_CancelHoldScrolling();
+            } else {
```
Comment (Brollyy) [inline Windows/MainWindow.GridControls.cs:194]: This should probably be outside of else - check possible problems when transitioning from scrolling to selecting with the same left click.

File: Windows/MainWindow.GridControls.cs [hunk 3]
```diff
@@ -135,24 +189,38 @@ public partial class MainWindow : Window {
...
             if (!editorIsLoaded) {
                 return;
             }
-            Point mousePos = e.GetPosition(EditorGrid);
-            gridController.GridMouseUp(mousePos);
-            editorMouseDown = false;
+            if (editorIsHoldScrolling) { //Moving with a released wheel and pressing a button
+                scrollEditor_CancelHoldScrolling();
+            }
+            if (editorDragSelectStart.HasValue || !editorIsHoldScrolling) {
```
Comment (Brollyy) [inline Windows/MainWindow.GridControls.cs:208]: `scrollEditor_CancelHoldScrolling` will update `editorIsHoldScrolling`, check if this logic behaves as expected.

File: Windows/MainWindow.GridControls.cs (hunk 2 above)
Comment (Brollyy) [reply Windows/MainWindow.GridControls.cs:194]: After more testing, this is fine as it is - allowing this logic to fire when hold scrolling ends causes unintuitive behavior with left click, which would require a major refactor to solve.

**Review by PKBeam — APPROVED** [review]

**Pull request merged**